util/tls/tls_socket.cc

```diff
@@ -488,8 +521,135 @@ io::Result<size_t> TlsSocket::TrySend(const iovec* v, uint32_t len) {
+    // error / protocol violation
+    if (read_result > 0) {  // case 1:
+      buf.remove_prefix(read_result);
+      total_bytes_read += read_result;
+      continue;
+    } else if ((read_result == Engine::NEED_READ_AND_MAYBE_WRITE) ||
+               (read_result == Engine::NEED_WRITE)) {  // case 2: Handle NEED_READ/WRITE
+      ///////////////////////////////////////////////////////////////
+      // 1. Check for a write conflict:
+      // Write conflicts are expected: The TLS state machine (e.g., renegotiation) may trigger
+      // writes during a read operation.
+      if (write_in_progress) {
+        // Another fiber is handling TLS writes (handshake/renegotiation);
+        // we cannot safely proceed without blocking this read fiber.
+        // Return partial data if available, or signal retry (EAGAIN) to caller.
+        if (total_bytes_read > 0) {
+          break;
+        }
+        return make_unexpected(make_error_code(errc::resource_unavailable_try_again));
+      }
+      ///////////////////////////////////////////////////////////////
+
+      // 2. Handle Pending Output (write_in_progress is false)
+      // If the engine generated TLS data (handshake/alerts), flush it now.
+      // Otherwise, skip to reading.
+      size_t output_pending_bytes{engine_->OutputPending()};
+      if ((read_result == Engine::NEED_WRITE) && (output_pending_bytes == 0)) {  // sanity check
+        // Critical Error: Engine demands a write but provided no data.
+        // Falling through to read would likely result in a stall/deadlock.
+        LOG(DFATAL) << "SSL BUG: Engine returned NEED_WRITE but OutputPending is 0";
+        return make_unexpected(make_error_code(errc::protocol_error));
+      }
+
+      if (output_pending_bytes > 0) {
+        auto output_buf{engine_->PeekOutputBuf()};
+        auto send_result{next_sock_->TrySend(output_buf)};
+        if (!send_result) {
+          // Write failed (EAGAIN or Error).
+          if (total_bytes_read > 0)
+            break;
+          return make_unexpected(send_result.error());
+        }
+        engine_->ConsumeOutputBuf(*send_result);
+
+        // If we couldn't send everything, we can't proceed safely.
+        if ((*send_result) < output_buf.size()) {
+          if (total_bytes_read > 0)
+            break;
+          return make_unexpected(make_error_code(errc::resource_unavailable_try_again));
+        }
+
+        // If we handled a NEED_WRITE successfully, we loop again to see what engine wants next
+        if (read_result == Engine::NEED_WRITE)
+          continue;
+      }
+      ///////////////////////////////////////////////////////////////
+      // 3. Check for read conflict:
+      // A read conflict implies the application is polling TryRecv while concurrently
+      // blocked on Recv, which is a usage error. We check this to prevent buffer corruption/crash.
+      if (read_in_progress) {
+        if (total_bytes_read > 0)
```

Comment:
you have this repeating pattern of 
```
if (total >0) 
  break;
 return ...
```

but you can just break, and then at line 652, check the status of total and either return total or the error.